File: db4o.net/Db4objects.Db4o/Db4objects.Db4o/Internal/Weakref/EnabledWeakReferenceSupport.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Internal.Weakref;
using Sharpen.Lang;

namespace Db4objects.Db4o.Internal.Weakref
{
	internal class EnabledWeakReferenceSupport : IWeakReferenceSupport
	{
		private readonly object _queue;

		private readonly ObjectContainerBase _container;

		private SimpleTimer _timer;

		internal EnabledWeakReferenceSupport(ObjectContainerBase container)
		{
			_container = container;
			_queue = Platform4.CreateReferenceQueue();
		}

		public virtual object NewWeakReference(ObjectReference referent, object obj)
		{
			return Platform4.CreateActiveObjectReference(_queue, referent, obj);
		}

		public virtual void Purge()
		{
			Platform4.PollReferenceQueue(_container, _queue);
		}

		public virtual void Start()
		{
			if (_timer != null)
			{
				return;
			}
			if (!_container.ConfigImpl.WeakReferences())
			{
				return;
			}
			if (_container.ConfigImpl.WeakReferenceCollectionInterval() <= 0)
			{
				return;
			}
			_timer = new SimpleTimer(new EnabledWeakReferenceSupport.Collector(this), _container
				.ConfigImpl.WeakReferenceCollectionInterval());
			_container.ThreadPool().Start("db4o WeakReference collector", _timer);
		}

		public virtual void Stop()
		{
			if (_timer == null)
			{
				return;
			}
			_timer.Stop();
			_timer = null;
		}

		private sealed class Collector : IRunnable
		{
			public void Run()
			{
				try
				{
					this._enclosing.Purge();
				}
				catch (DatabaseClosedException)
				{
				}
				catch (Exception e)
				{
					// can happen, no stack trace
					// don't bring down the thread
					Sharpen.Runtime.PrintStackTrace(e);
				}
			}

			internal Collector(EnabledWeakReferenceSupport _enclosing)
			{
				this._enclosing = _enclosing;
			}

			private readonly EnabledWeakReferenceSupport _enclosing;
		}
	}
}

File: db4oj/core/src/com/db4o/internal/weakref/EnabledWeakReferenceSupport.java
/* Copyright (C) 2004   Versant Inc.   http://www.db4o.com */

package com.db4o.internal.weakref;

import com.db4o.ext.*;
import com.db4o.foundation.*;
import com.db4o.internal.*;

class EnabledWeakReferenceSupport implements WeakReferenceSupport {
   
	private final Object _queue;
    private final ObjectContainerBase _container;
    private SimpleTimer _timer;
    
    EnabledWeakReferenceSupport(ObjectContainerBase container) {
        _container = container;
        _queue = Platform4.createReferenceQueue();
    }

    public Object newWeakReference(ObjectReference referent, Object obj) {
        return Platform4.createActiveObjectReference(_queue, referent, obj);
    }

    public void purge() {
        Platform4.pollReferenceQueue(_container, _queue);
    }
    
    public void start() {
    	if (_timer != null) {
    		return;
    	}
    	
        if(! _container.configImpl().weakReferences()){
            return;
        }
    	
        if (_container.configImpl().weakReferenceCollectionInterval() <= 0) {
        	return;
        }
        
        _timer = new SimpleTimer(new Collector(), _container.configImpl().weakReferenceCollectionInterval());
        _container.threadPool().start("db4o WeakReference collector", _timer);
    }

    /* (non-Javadoc)
	 * @see com.db4o.internal.WeakReferenceSupport#stopTimer()
	 */
    public void stop() {
    	if (_timer == null){
            return;
        }
        _timer.stop();
        _timer = null;
    }
    
    private final class Collector implements Runnable {
		public void run() {
			try {
				purge();
			} catch (DatabaseClosedException dce) {
				// can happen, no stack trace
			} catch (Exception e) {
				// don't bring down the thread
				e.printStackTrace();
			}
		}
	}

    
}